### PR TITLE
Upgrade Go to 1.6.

### PR DESCRIPTION
**Pull request by calavera**
I'm doing one Dockerfile at a time, just to get a feel of it.

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

File: opts/hosts_test.go [hunk 1]
```diff
@@ -8,7 +8,7 @@ import (
 func TestParseHost(t *testing.T) {
 	invalid := map[string]string{
 		"anything":              "Invalid bind address format: anything",
-		"something with spaces": "Invalid bind address format: something with spaces",
+		"something with spaces": "parse tcp://something with spaces: invalid character \" \" in host name",
```
Comment (thaJeztah) [inline opts/hosts_test.go]: Hm, unrelated change?

(Reviewing one PR at a time, just to get a hang of it)

Comment (calavera) [inline opts/hosts_test.go]: nope, they really change that message :pensive: Before:

```
FAIL: TestParseHost (0.00s)
	hosts_test.go:42: Expected an error for something with spaces with [Invalid bind address format: something with spaces], got [parse tcp://something with spaces: invalid character " " in host name]
```

You can see that janky is not failing in that test.

Comment (thaJeztah) [inline opts/hosts_test.go]: Oh! sorry for the noise then; I thought it wasn't related to this change :disappointed:

**Comment by aaronlehmann** [comment]
Interesting that the trust tests are failing after the Go upgrade. I looked at the tuf code that's throwing this error and I don't see anything obvious. @endophage any ideas?

**Comment by endophage** [comment]
@aaronlehmann seeing if we can repo in notary.

**Comment by endophage** [comment]
Go 1.6 is producing errors in Notary's own test suite. Working on tracking down the problems now

**Comment by endophage** [comment]
Tracked the trust errors down to a change in JSON (de)serialization logic loosely described here: https://golang.org/doc/go1.6#reflect

We had a type `tufKey` that contained our common key fields and was embedded into each of our typed key structs. Prior to Go 1.6, all embedded structs would be traversed during JSON (de)serialization. Now they follow more traditional export rules so only embedded structs that are exported will be traversed. This was causing JSON to strip out the public keys. Renaming the type to `TUFKey` solves the problem. Will be submitting a vendored minor patch to fix so you can make progress here.

**Comment by calavera** [comment]
> We had a type tufKey that contained our common key fields and was embedded into each of our typed key structs. Prior to Go 1.6, all embedded structs would be traversed during JSON (de)serialization.

that makes me wonder if we have more cases like that. Thanks for tracking this down @endophage.

File: opts/hosts_test.go (hunk 1 above)
Comment (MHBauer) [inline opts/hosts_test.go]: looks like https://github.com/golang/go/commit/617c93ce740c3c3cc28cdd1a0d712be183d0b328 did it

**Comment by aaronlehmann** [comment]
BTW a new version of notary was vendored. Rerunning the tests to see if they will pass now.

**Comment by icecrime** [comment]
CI is green :green_heart:

**Comment by cpuguy83** [comment]
woah

**Comment by LK4D4** [comment]
OMG
LGTM
it's time for new vendor also

**Comment by vdemeester** [comment]
LGTM 🐍

/cc @docker/core-engine-maintainers for review as well, just to make sure we're all ok with it and not scared :scream_cat:

File: Dockerfile [hunk 2]
```diff
@@ -116,7 +116,7 @@ RUN set -x \
 # IMPORTANT: If the version of Go is updated, the Windows to Linux CI machines
 #            will need updating, to avoid errors. Ping #docker-maintainers on IRC
 #            with a heads-up.
-ENV GO_VERSION 1.5.3
+ENV GO_VERSION 1.6
```
Comment (vdemeester) [inline Dockerfile]: `1.6` vs `1.6.0` ? :angel:

**Comment by thaJeztah** [comment]
LGTM - just discussed with @vdemeester and it's better to find possibly issues with 1.6 now, then later, so :+1:

**Comment by unclejack** [comment]
LGTM

**Comment by coolljt0725** [comment]
LGTM

**Comment by unclejack** [comment]
It seems we have enough LGTMs. I'll merge this.

**Comment by thaJeztah** [comment]
balancing-on-the-edge
